#964 Prompt move sketch when opening invalid sketch outside from IDE2

kittaakos opened the pull request:
### Motivation
<!-- Why this pull request? -->

The sketch folder name must match the name of the main sketch file. See the [specs](https://arduino.github.io/arduino-cli/dev/sketch-specification/#sketch-root-folder). This PR makes IDE2 more resilient by offering a sketch move to the user. This PR proposes the same behavior as IDE 1.x.

### Change description
<!-- What does your code do? -->

It's not much change. IDE2 tries to parse the error message from the CLI and guess an invalid sketch folder name from the error message. The rest of the logic already existed;  IDE2 does the same as `File` > `Open` > and tries to open an invalid sketch by prompting a move.

### Other information
<!-- Any additional information that could help the review process -->

Closes #964

### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

per1234 reviewed (changes requested):
Dialog is not shown when opening sketch with mismatched file/folder names on Windows
### To reproduce

Follow the instructions for (9) above.

🐛 The IDE starts up with a "fallback sketch" without showing a dialog to communicate the problem with the sketch to the user.

### Expected behavior

The same behavior as when I attempt to open the file via **File > Open**:

![image](https://user-images.githubusercontent.com/8572152/196900021-f0681fa0-148e-4638-b80b-b3a915adc1c5.png)

### Arduino IDE version

2.0.1-snapshot-c2f6cd6 (tester build for https://github.com/arduino/arduino-ide/pull/1563/commits/37c30d5eae4c4b157cda95119fd2a886e96d3865)

### Operating system

Windows

### Operating system version

10

### Additional context

The IDE is working as expected under these conditions on Linux.

---

The problem occurs with both absolute and relative sketch paths.

The problem occurs with paths to both the sketch folder and sketch file.

---

I see this in the daemon logs (in this case I opened the sketch file via file association because the daemon logs output hangs the command line terminal):

```text
2022-10-20 01:26:51 daemon INFO 4 CALLED: /cc.arduino.cli.commands.v1.ArduinoCoreService/LoadSketch

2022-10-20 01:26:51 daemon INFO 4 |  REQ:  {
4 |    "instance": {
4 |      "id": 1
4 |    },
4 |    "sketch_path": "c:\\ide 2\\rev\\untitled folder2"
4 |  }

2022-10-20 01:26:51 daemon INFO 4 |  ERROR:  rpc error: code = NotFound desc = Can't open sketch: main file missing from sketch: C:\ide 2\rev\untitled folder2\untitled folder2.ino

2022-10-20 01:26:51 daemon INFO 4 |  RESP: null
4 CALL END

5 CALLED: /cc.arduino.cli.commands.v1.ArduinoCoreService/LoadSketch
5 |  REQ:  {
5 |    "instance": {
5 |      "id": 1
5 |    },
5 |    "sketch_path": "c:\\ide 2\\rev"
5 |  }

2022-10-20 01:26:51 daemon INFO 5 |  ERROR:  rpc error: code = NotFound desc = Can't open sketch: main file missing from sketch: C:\ide 2\rev\rev.ino
5 |  RESP: null
5 CALL END

6 CALLED: /cc.arduino.cli.commands.v1.ArduinoCoreService/LoadSketch
6 |  REQ:  {
6 |    "instance": {
6 |      "id": 1
6 |    },
6 |    "sketch_path": "c:\\ide 2"
6 |  }

2022-10-20 01:26:51 daemon INFO 6 |  ERROR:  rpc error: code = NotFound desc = Can't open sketch: main file missing from sketch: C:\ide 2\ide 2.ino
6 |  RESP: null
6 CALL END
```

The first request matches what I see on the Linux machine, but the next two where the IDE attempts to load the parent paths as sketches is unique and strange.

### Issue checklist

- [X] I searched for previous reports in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the problem still occurs when using the latest [nightly build](https://github.com/arduino/arduino-ide#nightly-builds)
- [X] My report contains all necessary details

AlbyIanna reviewed (changes requested):
I tested all the cases on MacOS, and here are the results:
1. ❌
**expected:** IDE2 opens and reloads with a fallback sketch. No notifications, no dialogs. A temp sketch is opened with the current date.
**actual:** IDE2 opens the last open sketch (not a temp sektch)
2. ✅ 
3. ✅
4. ✅
5. ✅
6. ✅
7. ✅
8. ✅
9. ❌ - I think there's just an error in the description of step 9, because in the premise you say to open `fails_to_open.ino`:

> Open the untitled folder2/fails_to_open.ino main sketch file from a terminal.
> 
> `./Arduino\ IDE.app/Contents/MacOS/Arduino\ IDE ./untitled\ folder2/fails_to_open.ino`

and then you refer to `sketch.ino`:
> Expectations are the same as for step 8. Click on OK. IDE2 reloads and moves the sketch to `./untitled folder2/sketch/sketch.ino`. IDE2 is functional.

Probably you wanted to write:
> 9. (corrected) Open the untitled folder2/sketch.ino main sketch file from a terminal.
> 
> `./Arduino\ IDE.app/Contents/MacOS/Arduino\ IDE ./untitled\ folder2/sketch.ino`
>
> Expectations are the same as for step 8. Click on OK. IDE2 reloads and moves the sketch to `./untitled folder2/sketch/sketch.ino`. IDE2 is functional.

Am I right? In that case,
9. (corrected) ✅ 
10. ✅



## Arduino IDE version
12.0.1-snapshot-c2f6cd6 (tester build for https://github.com/arduino/arduino-ide/commit/37c30d5eae4c4b157cda95119fd2a886e96d3865)

## Operating system
MacOS

## Operating system version
13.3.1 (21E258)

kittaakos commented:
> Am I right? In that case,

Yes, you're right. Thank you!

kittaakos commented:
Great review 👍 

> * ❌
>   **expected:** IDE2 opens and reloads with a fallback sketch. No notifications, no dialogs. A temp sketch is opened with the current date.
>   **actual:** IDE2 opens the last open sketch (not a temp sektch)

This has been fixed. ✅ 

Added additional notes about the expected behavior to the PR [description](https://github.com/arduino/arduino-ide/pull/1563#issue-1411912180): 👇 


> When the IDE2 is started from a terminal, the `path` input can be a folder or a file:
> 
> * if folder:
>   
>   * if exists and contains a valid main sketch file, the `path` will be used,
>   * if exists but contains only invalid (main) sketch files, the `path` will be used,
>   * if exists but does not contain any sketch files, the `path` will be used, but IDE2 will reload and will provide a fallback sketch,
>   * if missing, IDE2 ignores the `path` argument and restores the previous sketches, if any. Otherwise, a fallback sketch will open.
> * if file:
>   
>   * if exists and is a `.ino` file, the parent folder will be used as the `path` to open the sketch,
>   * if missing or not a `.ino` file, IDE2 ignores the `path` argument and restores the previous sketches, if any. Otherwise, a fallback sketch will open.


> 9\. ❌ - I think there's just an error in the description of step 9, because in the premise you say to open `fails_to_open.ino`:

Correct. The error was in the description. I have updated the expectation. ✅ 

I have downloaded the test build and went through the steps. All worked on my macOS installation. Please take another look, and let me know how it works. Thank you! 🙏

AlbyIanna commented:
I did another run of tests with a new build, now **1.** is working, but **7.** fails.

Here are the new results
1. ✅
2. ✅ 
3. ✅
4. ✅
5. ✅
6. ✅
7. ❌

> Open IDE2 by clicking on the untitled folder2/fails_to_open.ino icon. (See step 6. on how to open.)
> 
> A fallback sketch opens. A modal dialog says that the fails_to_open.ino must be in a correctly named sketch folder. There is a notification message showing the original error message from the CLI.

**Result**: IDE2 opens and reloads with a fallback sketch. No notifications, no dialogs. A temp sketch is opened with the current date.
<details>
  <summary><b>Screen recording</b></summary>
  
https://user-images.githubusercontent.com/6939054/196964325-a80573d1-d02b-45a4-a0e5-1553e004a4c2.mov

</details>
8. ✅
9. ✅ 
10. ✅



## Arduino IDE version
12.0.1-snapshot-de4e865 (tester build for https://github.com/arduino/arduino-ide/commit/91c7cef315bbb9dfbd484b6a49365ec325fb0a42)

## Operating system
MacOS

## Operating system version
13.3.1 (21E258)

kittaakos commented:
> IDE2 shows the error from the CLI:
> 
> <img alt="Screen Shot 2022-10-19 at 18 59 07" width="525" src="https://user-images.githubusercontent.com/1405703/196756479-284e093e-24ed-4f7c-9a34-367c576882f8.png">
> 
> * IDE2 can omit to show anything,
> * Can tweak the error message,
> * CLI can provide a more user-friendly message.

I have discussed this with Cristian, and the CLI can provide a fine-grained error message if the sketch (folder) name is invalid, etc. This will come later. For the time being, IDE2 will show the original error message from the CLI.

kittaakos commented:
> but **7.** fails.

I could not reproduce it. It works here with the build. Could you please double-check if you have picked the correct IDE2 version to open the invalid sketch?


https://user-images.githubusercontent.com/1405703/196974538-ef641b8f-b2be-44f8-a681-627049b13db1.mp4

kittaakos commented:
> if you have picked the correct IDE2 version



> **Result**: IDE2 opens and reloads with a fallback sketch. No notifications, no dialogs. A temp sketch is opened with the current date.

The attached video helped a lot. Yes, you opened the default IDE2, a nightly build by double-clicking on the `.ino` file. See the window title.

AlbyIanna commented:
> The attached video helped a lot. Yes, you opened the default IDE2, a nightly build by double-clicking on the .ino file. See the window title.

oh, you're right! Sorry for that!
I tested it again with right build and it works correctly!

AlbyIanna reviewed (approved):
Everything's working as expected for me.
Also code, looks good to me (I've just left two unimportant remarks)

Thank you Akos!

kittaakos commented:
@per1234, please take another look if you have time. Thank you!

per1234 reviewed (changes requested):
I still experience the same problem as [before](https://github.com/arduino/arduino-ide/pull/1563#pullrequestreview-1148802369):

Dialog is not shown when opening sketch with mismatched file/folder names on Windows
### To reproduce

1. Create a sketch which does not contain a `.ino` file matching the folder name:
   ```text
   Foo/
   └── Bar.ino
   ```
1. Open the Arduino IDE installation folder in a command line terminal.
1. Pass the path to the sketch with the file/folder name mismatch as an argument to the Arduino IDE invocation:
   ```text
   & ".\Arduino IDE" "C:\Users\per\Documents\Arduino\Foo\Bar.ino"
   ```
1. Wait for the Arduino IDE to finish starting up.

🐛 The IDE starts up with a "fallback sketch" without showing a dialog to communicate the problem with the sketch to the user.

### Expected behavior

The same behavior as when I attempt to open the file via **File > Open**:

![image](https://user-images.githubusercontent.com/8572152/197849456-080a5566-5cdb-427b-b554-1a23ee1be8cc.png)

### Arduino IDE version

2.0.1-snapshot-6324f75 (tester build for https://github.com/arduino/arduino-ide/pull/1563/commits/fc4a668ce8554ffb9d9247fd80d904c66e0bf802)

### Operating system

Windows

### Operating system version

10

### Additional context

The IDE is working as expected under these conditions on Linux.

---

The problem occurs with both absolute and relative sketch paths.

The problem occurs with paths to both the sketch folder and sketch file.

kittaakos commented:
I could reproduce the defect on Windows and fix it. I have verified the case mentioned in https://github.com/arduino/arduino-ide/pull/1563#pullrequestreview-1155289165, and it worked.

Please review. Thank you!

per1234 reviewed (approved):
It is working perfectly for me now. I verified it:

- fixes https://github.com/arduino/arduino-ide/issues/964
- resolves https://github.com/arduino/arduino-ide/issues/1484
- adds support for specifying path to sketch folder instead of file
- adds support for relative sketch paths

Thanks Akos!
